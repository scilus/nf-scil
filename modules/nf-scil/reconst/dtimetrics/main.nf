
process RECONST_DTIMETRICS {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
        tuple val(meta), path(dwi), path(bval), path(bvec), path(b0mask)

    output:
        tuple val(meta), path("*__ad.nii.gz")                      , emit: ad, optional: true
        tuple val(meta), path("*__evecs.nii.gz")                   , emit: evecs, optional: true
        tuple val(meta), path("*__evecs_v1.nii.gz")                , emit: evecs_v1, optional: true
        tuple val(meta), path("*__evecs_v2.nii.gz")                , emit: evecs_v2, optional: true
        tuple val(meta), path("*__evecs_v3.nii.gz")                , emit: evecs_v3, optional: true
        tuple val(meta), path("*__evals.nii.gz")                   , emit: evals, optional: true
        tuple val(meta), path("*__evals_e1.nii.gz")                , emit: evals_e1, optional: true
        tuple val(meta), path("*__evals_e2.nii.gz")                , emit: evals_e2, optional: true
        tuple val(meta), path("*__evals_e3.nii.gz")                , emit: evals_e3, optional: true
        tuple val(meta), path("*__fa.nii.gz")                      , emit: fa, optional: true
        tuple val(meta), path("*__ga.nii.gz")                      , emit: ga, optional: true
        tuple val(meta), path("*__rgb.nii.gz")                     , emit: rgb, optional: true
        tuple val(meta), path("*__md.nii.gz")                      , emit: md, optional: true
        tuple val(meta), path("*__mode.nii.gz")                    , emit: mode, optional: true
        tuple val(meta), path("*__norm.nii.gz")                    , emit: norm, optional: true
        tuple val(meta), path("*__rd.nii.gz")                      , emit: rd, optional: true
        tuple val(meta), path("*__tensor.nii.gz")                  , emit: tensor, optional: true
        tuple val(meta), path("*__nonphysical.nii.gz")             , emit: nonphysical, optional: true
        tuple val(meta), path("*__pulsation_std_dwi.nii.gz")       , emit: pulsation_std_dwi, optional: true
        tuple val(meta), path("*__residual.nii.gz")                , emit: residual, optional: true
        tuple val(meta), path("*__residual_iqr_residuals.npy")     , emit: residual_iqr_residuals, optional: true
        tuple val(meta), path("*__residual_mean_residuals.npy")    , emit: residual_mean_residuals, optional: true
        tuple val(meta), path("*__residual_q1_residuals.npy")      , emit: residual_q1_residuals, optional: true
        tuple val(meta), path("*__residual_q3_residuals.npy")      , emit: residual_q3_residuals, optional: true
        tuple val(meta), path("*__residual_residuals_stats.png")   , emit: residual_residuals_stats, optional: true
        tuple val(meta), path("*__residual_std_residuals.npy")     , emit: residual_std_residuals, optional: true
        path "versions.yml"                                        , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def dwi_shell_tolerance = task.ext.dwi_shell_tolerance ? "--tolerance " + task.ext.dwi_shell_tolerance : ""
    def max_dti_shell_value = task.ext.max_dti_shell_value ?: 1500
    def b0_thr_extract_b0 = task.ext.b0_thr_extract_b0 ?: 10
    def b0_threshold = task.ext.b0_thr_extract_b0 ? "--b0_threshold $task.ext.b0_thr_extract_b0" : ""
    def dti_shells = task.ext.dti_shells ?: "\$(cut -d ' ' --output-delimiter=\$'\\n' -f 1- $bval | awk -F' ' '{v=int(\$1)}{if(v<=$max_dti_shell_value|| v<=$b0_thr_extract_b0)print v}' | uniq)"

    if ( b0mask ) args += " --mask $b0mask"
    if ( task.ext.ad ) args += " --ad ${prefix}__ad.nii.gz"
    if ( task.ext.evecs ) args += " --evecs ${prefix}__evecs.nii.gz"
    if ( task.ext.evals ) args += " --evals ${prefix}__evals.nii.gz"
    if ( task.ext.fa ) args += " --fa ${prefix}__fa.nii.gz"
    if ( task.ext.ga ) args += " --ga ${prefix}__ga.nii.gz"
    if ( task.ext.rgb ) args += " --rgb ${prefix}__rgb.nii.gz"
    if ( task.ext.md ) args += " --md ${prefix}__md.nii.gz"
    if ( task.ext.mode ) args += " --mode ${prefix}__mode.nii.gz"
    if ( task.ext.norm ) args += " --norm ${prefix}__norm.nii.gz"
    if ( task.ext.rd ) args += " --rd ${prefix}__rd.nii.gz"
    if ( task.ext.tensor ) args += " --tensor ${prefix}__tensor.nii.gz"
    if ( task.ext.nonphysical ) args += " --non-physical ${prefix}__nonphysical.nii.gz"
    if ( task.ext.pulsation ) args += " --pulsation ${prefix}__pulsation_std_dwi.nii.gz"
    if ( task.ext.residual ) args += " --residual ${prefix}__residual.nii.gz"


    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_dwi_extract_shell.py $dwi $bval $bvec $dti_shells \
                dwi_dti_shells.nii.gz bval_dti_shells bvec_dti_shells \
                $dwi_shell_tolerance -f

    scil_dti_metrics.py dwi_dti_shells.nii.gz bval_dti_shells bvec_dti_shells \
        --not_all $args $b0_threshold -f

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_dwi_extract_shell.py -h
    scil_dti_metrics.py -h

    touch ${prefix}__ad.nii.gz
    touch ${prefix}__evecs.nii.gz
    touch ${prefix}__evecs_v1.nii.gz
    touch ${prefix}__evecs_v2.nii.gz
    touch ${prefix}__evecs_v3.nii.gz
    touch ${prefix}__evals.nii.gz
    touch ${prefix}__evals_e1.nii.gz
    touch ${prefix}__evals_e2.nii.gz
    touch ${prefix}__evals_e3.nii.gz
    touch ${prefix}__fa.nii.gz
    touch ${prefix}__ga.nii.gz
    touch ${prefix}__rgb.nii.gz
    touch ${prefix}__md.nii.gz
    touch ${prefix}__mode.nii.gz
    touch ${prefix}__norm.nii.gz
    touch ${prefix}__rd.nii.gz
    touch ${prefix}__tensor.nii.gz
    touch ${prefix}__nonphysical.nii.gz
    touch ${prefix}__pulsation_std_dwi.nii.gz
    touch ${prefix}__residual.nii.gz
    touch ${prefix}__residual_iqr_residuals.npy
    touch ${prefix}__residual_mean_residuals.npy
    touch ${prefix}__residual_q1_residuals.npy
    touch ${prefix}__residual_q3_residuals.npy
    touch ${prefix}__residual_residuals_stats.png
    touch ${prefix}__residual_std_residuals.npy

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """
}
