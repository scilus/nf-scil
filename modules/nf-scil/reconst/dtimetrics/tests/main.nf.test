nextflow_process {

    name "Test Process RECONST_DTIMETRICS"
    script "../main.nf"
    process "RECONST_DTIMETRICS"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "reconst"
    tag "reconst/dtimetrics"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "heavy.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("reconst - dtimetrics") {
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi/dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/dwi/dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/dwi/dwi.bvec", checkIfExists: true),
                        []
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("reconst - dtimetrics_with_b0mask") {
        config "./nextflow_light.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi/dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/dwi/dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/dwi/dwi.bvec", checkIfExists: true),
                        file("\${test_data_directory}/dwi/mask.nii.gz", checkIfExists: true)
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
