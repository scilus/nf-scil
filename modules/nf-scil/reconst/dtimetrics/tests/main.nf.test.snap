{
    "reconst - dtimetrics_with_b0mask": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "10": [
                    
                ],
                "11": [
                    
                ],
                "12": [
                    
                ],
                "13": [
                    
                ],
                "14": [
                    
                ],
                "15": [
                    
                ],
                "16": [
                    
                ],
                "17": [
                    
                ],
                "18": [
                    
                ],
                "19": [
                    
                ],
                "2": [
                    
                ],
                "20": [
                    
                ],
                "21": [
                    
                ],
                "22": [
                    
                ],
                "23": [
                    
                ],
                "24": [
                    
                ],
                "25": [
                    
                ],
                "26": [
                    "versions.yml:md5,609a8b62c7dc862692683ec5860ba677"
                ],
                "3": [
                    
                ],
                "4": [
                    
                ],
                "5": [
                    
                ],
                "6": [
                    
                ],
                "7": [
                    
                ],
                "8": [
                    
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fa.nii.gz:md5,2e52b551402bdbdec9a301b875fc4722"
                    ]
                ],
                "ad": [
                    
                ],
                "evals": [
                    
                ],
                "evals_e1": [
                    
                ],
                "evals_e2": [
                    
                ],
                "evals_e3": [
                    
                ],
                "evecs": [
                    
                ],
                "evecs_v1": [
                    
                ],
                "evecs_v2": [
                    
                ],
                "evecs_v3": [
                    
                ],
                "fa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fa.nii.gz:md5,2e52b551402bdbdec9a301b875fc4722"
                    ]
                ],
                "ga": [
                    
                ],
                "md": [
                    
                ],
                "mode": [
                    
                ],
                "nonphysical": [
                    
                ],
                "norm": [
                    
                ],
                "pulsation_std_dwi": [
                    
                ],
                "rd": [
                    
                ],
                "residual": [
                    
                ],
                "residual_iqr_residuals": [
                    
                ],
                "residual_mean_residuals": [
                    
                ],
                "residual_q1_residuals": [
                    
                ],
                "residual_q3_residuals": [
                    
                ],
                "residual_residuals_stats": [
                    
                ],
                "residual_std_residuals": [
                    
                ],
                "rgb": [
                    
                ],
                "tensor": [
                    
                ],
                "versions": [
                    "versions.yml:md5,609a8b62c7dc862692683ec5860ba677"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-20T23:06:15.147391"
    },
    "reconst - dtimetrics": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ad.nii.gz:md5,ece4e1d0fc03f4bafc882b63285984d3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs.nii.gz:md5,15ccf781a3329ea61e12ded2e58db397"
                    ]
                ],
                "10": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ga.nii.gz:md5,b65ebb5bb38a3e47e83ea0a569e9e4b1"
                    ]
                ],
                "11": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__rgb.nii.gz:md5,601b67930e79dbf23dfaeb5aafe6f3e7"
                    ]
                ],
                "12": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__md.nii.gz:md5,d9a6ca6d7ad205b8ec0a90f0db4c53d9"
                    ]
                ],
                "13": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mode.nii.gz:md5,4b964b1b2eaa376e138d956822e6673f"
                    ]
                ],
                "14": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__norm.nii.gz:md5,5540436e8e61effe47dde7a1c89cfb9a"
                    ]
                ],
                "15": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__rd.nii.gz:md5,d33a7089844a53184c0548cdf8899efa"
                    ]
                ],
                "16": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__tensor.nii.gz:md5,5d74f987559fa980fd899d26c4b6f101"
                    ]
                ],
                "17": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nonphysical.nii.gz:md5,7da6de5ad6e7c37518f9f3962f066f89"
                    ]
                ],
                "18": [
                    
                ],
                "19": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual.nii.gz:md5,72090f28a6d0aaf0654def397f59d2ba"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs_v1.nii.gz:md5,2ee5bccdf2a591d2b6d6f2b7067d05e4"
                    ]
                ],
                "20": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_iqr_residuals.npy:md5,4a71d46303bb3ac67088fe5cdaa5694a"
                    ]
                ],
                "21": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_mean_residuals.npy:md5,2247858eaf51c6e973649418f8721477"
                    ]
                ],
                "22": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_q1_residuals.npy:md5,4a71d46303bb3ac67088fe5cdaa5694a"
                    ]
                ],
                "23": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_q3_residuals.npy:md5,4a71d46303bb3ac67088fe5cdaa5694a"
                    ]
                ],
                "24": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_residuals_stats.png:md5,18ce14e9e554c8f5fb1cf0c87af36f9f"
                    ]
                ],
                "25": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_std_residuals.npy:md5,5e526c12b8a13b611bd1b2e00d4e1448"
                    ]
                ],
                "26": [
                    "versions.yml:md5,609a8b62c7dc862692683ec5860ba677"
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs_v2.nii.gz:md5,00cefe7aea0d8bb29bb00e29934d0085"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs_v3.nii.gz:md5,d8e9dce7cc788088110d40bb9dcee95b"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals.nii.gz:md5,62a56f74298f59038de8f3ac4e1d8f7e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals_e1.nii.gz:md5,ece4e1d0fc03f4bafc882b63285984d3"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals_e2.nii.gz:md5,eb0cba1c29152ad36a7768e977326829"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals_e3.nii.gz:md5,78744b78fa617c7c1a16ff824318359c"
                    ]
                ],
                "9": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fa.nii.gz:md5,d8a924bfab876fffca2173384c651ff3"
                    ]
                ],
                "ad": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ad.nii.gz:md5,ece4e1d0fc03f4bafc882b63285984d3"
                    ]
                ],
                "evals": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals.nii.gz:md5,62a56f74298f59038de8f3ac4e1d8f7e"
                    ]
                ],
                "evals_e1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals_e1.nii.gz:md5,ece4e1d0fc03f4bafc882b63285984d3"
                    ]
                ],
                "evals_e2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals_e2.nii.gz:md5,eb0cba1c29152ad36a7768e977326829"
                    ]
                ],
                "evals_e3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evals_e3.nii.gz:md5,78744b78fa617c7c1a16ff824318359c"
                    ]
                ],
                "evecs": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs.nii.gz:md5,15ccf781a3329ea61e12ded2e58db397"
                    ]
                ],
                "evecs_v1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs_v1.nii.gz:md5,2ee5bccdf2a591d2b6d6f2b7067d05e4"
                    ]
                ],
                "evecs_v2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs_v2.nii.gz:md5,00cefe7aea0d8bb29bb00e29934d0085"
                    ]
                ],
                "evecs_v3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__evecs_v3.nii.gz:md5,d8e9dce7cc788088110d40bb9dcee95b"
                    ]
                ],
                "fa": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__fa.nii.gz:md5,d8a924bfab876fffca2173384c651ff3"
                    ]
                ],
                "ga": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ga.nii.gz:md5,b65ebb5bb38a3e47e83ea0a569e9e4b1"
                    ]
                ],
                "md": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__md.nii.gz:md5,d9a6ca6d7ad205b8ec0a90f0db4c53d9"
                    ]
                ],
                "mode": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mode.nii.gz:md5,4b964b1b2eaa376e138d956822e6673f"
                    ]
                ],
                "nonphysical": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nonphysical.nii.gz:md5,7da6de5ad6e7c37518f9f3962f066f89"
                    ]
                ],
                "norm": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__norm.nii.gz:md5,5540436e8e61effe47dde7a1c89cfb9a"
                    ]
                ],
                "pulsation_std_dwi": [
                    
                ],
                "rd": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__rd.nii.gz:md5,d33a7089844a53184c0548cdf8899efa"
                    ]
                ],
                "residual": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual.nii.gz:md5,72090f28a6d0aaf0654def397f59d2ba"
                    ]
                ],
                "residual_iqr_residuals": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_iqr_residuals.npy:md5,4a71d46303bb3ac67088fe5cdaa5694a"
                    ]
                ],
                "residual_mean_residuals": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_mean_residuals.npy:md5,2247858eaf51c6e973649418f8721477"
                    ]
                ],
                "residual_q1_residuals": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_q1_residuals.npy:md5,4a71d46303bb3ac67088fe5cdaa5694a"
                    ]
                ],
                "residual_q3_residuals": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_q3_residuals.npy:md5,4a71d46303bb3ac67088fe5cdaa5694a"
                    ]
                ],
                "residual_residuals_stats": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_residuals_stats.png:md5,18ce14e9e554c8f5fb1cf0c87af36f9f"
                    ]
                ],
                "residual_std_residuals": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__residual_std_residuals.npy:md5,5e526c12b8a13b611bd1b2e00d4e1448"
                    ]
                ],
                "rgb": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__rgb.nii.gz:md5,601b67930e79dbf23dfaeb5aafe6f3e7"
                    ]
                ],
                "tensor": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__tensor.nii.gz:md5,5d74f987559fa980fd899d26c4b6f101"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,609a8b62c7dc862692683ec5860ba677"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-20T23:05:55.560305"
    }
}